src/main/java/co/cask/hydrator/python/transform/JythonPythonExecutor.java

```diff
@@ -0,0 +1,94 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.hydrator.python.transform;
+
+import co.cask.cdap.api.data.format.StructuredRecord;
+import co.cask.cdap.etl.api.Emitter;
+import co.cask.hydrator.common.script.ScriptContext;
+import org.python.core.Py;
+import org.python.core.PyCode;
+import org.python.core.PyException;
+import org.python.util.PythonInterpreter;
+
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.io.StringWriter;
+import java.util.Map;
+
+/**
+ * Executes python code using Jython library
+ */
+public class JythonPythonExecutor implements PythonExecutor {
+    private static final String INPUT_STRUCTURED_RECORD_VARIABLE_NAME = "dont_name_your_variable_this1";
+    private static final String EMITTER_VARIABLE_NAME = "dont_name_your_variable_this2";
+    private static final String CONTEXT_NAME = "dont_name_your_context_this";
+    private final PythonEvaluator.Config config;
+
+    private PythonInterpreter interpreter;
+    private PyCode compiledScript;
+
+    public JythonPythonExecutor(PythonEvaluator.Config config) {
+        this.config = config;
+    }
+
+    @Override
+    public void initialize(ScriptContext scriptContext) throws IOException, InterruptedException {
+        interpreter = new PythonInterpreter();
+        interpreter.set(CONTEXT_NAME, scriptContext);
+
+        // doing this so that we can pass the 'input' record into the transform function.
+        // that is, we want people to implement
+        // def transform(input, emitter, context): ...
+        // rather than def transform(): ...  and have them access the input, emitter, and context via global variables
+
+        String script = String.format("%s\ntransform(%s, %s, %s)",
+                config.getScript(), INPUT_STRUCTURED_RECORD_VARIABLE_NAME,
+                EMITTER_VARIABLE_NAME, CONTEXT_NAME);
+        compiledScript = interpreter.compile(script);
+    }
+
+    @Override
+    public void transform(StructuredRecord input, Emitter<StructuredRecord> emitter,
+                          Emitter<Map> pythonEmitter, ScriptContext scriptContext) {
+        try {
+            interpreter.set(INPUT_STRUCTURED_RECORD_VARIABLE_NAME,
+                    PythonObjectsEncoder.encode(input, input.getSchema()));
+            interpreter.set(EMITTER_VARIABLE_NAME, pythonEmitter);
+            Py.runCode(compiledScript, interpreter.getLocals(), interpreter.getLocals());
+
+        } catch (PyException e) {
+            // Put stack trace as the exception message, because otherwise the information from PyException is lost.
+            // PyException only exposes the actual cause (Python stack trace) if printStackTrace() is called on it.
+            throw new IllegalArgumentException("Could not transform input.\n" + getStackTrace(e));
```

Comment:
I still don't get it, why don't you pass it as a cause? Does CDAP logging system hide exception causes?

Reply:
I assume the exception is too big.
Here's what it shows if I remove getStackTrace method:

> co.cask.cdap.etl.batch.StageFailureException: Failed to execute pipeline stage 'Python' with the error: Could not transform input.
. Please review your pipeline configuration and check the system logs for more details.
	at co.cask.cdap.etl.batch.PipeStage.consume(PipeStage.java:52) ~[cdap-etl-batch-6.0.0-SNAPSHOT.jar:na]
	at co.cask.cdap.etl.batch.PipeEmitter.emit(PipeEmitter.java:83) ~[cdap-etl-batch-6.0.0-SNAPSHOT.jar:na]
	at co.cask.cdap.etl.common.TrackedEmitter.emit(TrackedEmitter.java:57) ~[cdap-etl-core-6.0.0-SNAPSHOT.jar:na]
	at co.cask.cdap.etl.common.plugin.UntimedEmitter.emit(UntimedEmitter.java:64) ~[cdap-etl-core-6.0.0-SNAPSHOT.jar:na]
	at co.cask.hydrator.plugin.batch.source.ExcelInputReader.transform(ExcelInputReader.java:205) ~[1549391341934-0/:na]
	at co.cask.cdap.etl.common.plugin.WrappedBatchSource.lambda$transform$2(WrappedBatchSource.java:69) [cdap-etl-core-6.0.0-SNAPSHOT.jar:na]
	at co.cask.cdap.etl.common.plugin.Caller$1.call(Caller.java:30) ~[cdap-etl-core-6.0.0-SNAPSHOT.jar:na]
	at co.cask.cdap.etl.common.plugin.StageLoggingCaller.call(StageLoggingCaller.java:40) ~[cdap-etl-core-6.0.0-SNAPSHOT.jar:na]
	at co.cask.cdap.etl.common.plugin.WrappedBatchSource.transform(WrappedBatchSource.java:68) [cdap-etl-core-6.0.0-SNAPSHOT.jar:na]
	at co.cask.cdap.etl.common.plugin.WrappedBatchSource.transform(WrappedBatchSource.java:36) [cdap-etl-core-6.0.0-SNAPSHOT.jar:na]
	at co.cask.cdap.etl.common.preview.LimitingTransform.transform(LimitingTransform.java:44) ~[cdap-etl-core-6.0.0-SNAPSHOT.jar:na]
	at co.cask.cdap.etl.common.TrackedTransform.transform(TrackedTransform.java:74) ~[cdap-etl-core-6.0.0-SNAPSHOT.jar:na]
	at co.cask.cdap.etl.batch.UnwrapPipeStage.consumeInput(UnwrapPipeStage.java:44) ~[cdap-etl-batch-6.0.0-SNAPSHOT.jar:na]
	at co.cask.cdap.etl.batch.UnwrapPipeStage.consumeInput(UnwrapPipeStage.java:32) ~[cdap-etl-batch-6.0.0-SNAPSHOT.jar:na]
	at co.cask.cdap.etl.batch.PipeStage.consume(PipeStage.java:44) ~[cdap-etl-batch-6.0.0-SNAPSHOT.jar:na]
	at co.cask.cdap.etl.batch.PipeTransformExecutor.runOneIteration(PipeTransformExecutor.java:43) ~[cdap-etl-batch-6.0.0-SNAPSHOT.jar:na]
	at co.cask.cdap.etl.batch.mapreduce.TransformRunner.transform(TransformRunner.java:142) ~[cdap-etl-batch-6.0.0-SNAPSHOT.jar:na]
	at co.cask.cdap.etl.batch.mapreduce.ETLMapReduce$ETLMapper.map(ETLMapReduce.java:465) ~[cdap-etl-batch-6.0.0-SNAPSHOT.jar:na]
	at org.apache.hadoop.mapreduce.Mapper.run(Mapper.java:146) ~[org.apache.hadoop.hadoop-mapreduce-client-core-2.8.0.jar:na]
	at co.cask.cdap.internal.app.runtime.batch.MapperWrapper.run(MapperWrapper.java:135) ~[na:na]
	at org.apache.hadoop.mapred.MapTask.runNewMapper(MapTask.java:787) ~[org.apache.hadoop.hadoop-mapreduce-client-core-2.8.0.jar:na]
	at org.apache.hadoop.mapred.MapTask.run(MapTask.java:341) ~[org.apache.hadoop.hadoop-mapreduce-client-core-2.8.0.jar:na]
	at org.apache.hadoop.mapred.LocalJobRunner$Job$MapTaskRunnable.run(LocalJobRunner.java:270) ~[org.apache.hadoop.hadoop-mapreduce-client-common-2.8.0.jar:na]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) ~[na:1.8.0_191]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[na:1.8.0_191]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) ~[na:1.8.0_191]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) ~[na:1.8.0_191]
	at java.lang.Thread.run(Thread.java:748) ~[na:1.8.0_191]
Caused by: java.lang.IllegalArgumentException: Could not transform input.
	at co.cask.hydrator.python.transform.JythonPythonExecutor.transform(JythonPythonExecutor.java:77) ~[1549391342283-0/:na]
	at co.cask.hydrator.python.transform.PythonEvaluator.transform(PythonEvaluator.java:233) ~[1549391342283-0/:na]
	at co.cask.hydrator.python.transform.PythonEvaluator.transform(PythonEvaluator.java:53) ~[1549391342283-0/:na]
	at co.cask.cdap.etl.common.plugin.WrappedTransform.lambda$transform$5(WrappedTransform.java:90) ~[cdap-etl-core-6.0.0-SNAPSHOT.jar:na]
	at co.cask.cdap.etl.common.plugin.Caller$1.call(Caller.java:30) ~[cdap-etl-core-6.0.0-SNAPSHOT.jar:na]
	at co.cask.cdap.etl.common.plugin.StageLoggingCaller.call(StageLoggingCaller.java:40) ~[cdap-etl-core-6.0.0-SNAPSHOT.jar:na]
	at co.cask.cdap.etl.common.plugin.WrappedTransform.transform(WrappedTransform.java:89) ~[cdap-etl-core-6.0.0-SNAPSHOT.jar:na]
	at co.cask.cdap.etl.common.TrackedTransform.transform(TrackedTransform.java:74) ~[cdap-etl-core-6.0.0-SNAPSHOT.jar:na]
	at co.cask.cdap.etl.batch.UnwrapPipeStage.consumeInput(UnwrapPipeStage.java:44) ~[cdap-etl-batch-6.0.0-SNAPSHOT.jar:na]
	at co.cask.cdap.etl.batch.UnwrapPipeStage.consumeInput(UnwrapPipeStage.java:32) ~[cdap-etl-batch-6.0.0-SNAPSHOT.jar:na]
	at co.cask.cdap.etl.batch.PipeStage.consume(PipeStage.java:44) ~[cdap-etl-batch-6.0.0-SNAPSHOT.jar:na]
	... 27 common frames omitted

Reply:
and how would same exception look like with getStackTrace? Strange that I cannot see Python code that causes an error.